### PR TITLE
[DPE-2639;DPE-2005] feat/test: kill, freeze + multiple cluser ha tests

### PR DESCRIPTION
**Pull request by marcoppenheimer**
Fixes https://github.com/canonical/zookeeper-k8s-operator/issues/42

## Review Notes
Code is **mostly** the same as in ZooKeeper VM.
This initial branch exists to add future PR's to with other HA tests.

Differences to VM (aka useful things to review):
- `helpers.get_unit_address_map` - A handier way of getting unit:host names. `jq` query slightly different as K8s fields a different
    - `helpers.get_hosts`, `helpers.get_unit_host` and `helpers.get_unit_name_from_host` all now use this method
- `helpers.send_control_signal` - For some unknown reason, `juju ssh`/`juju exec` both return non-0 exit codes
    - Worked around by using `microk8s.kubectl` to SSH in to the cluster and issue the `pkill` sigs
- `helpers.get_pid` - Addition useful for checking processes was killed and restarted (aka pid changed)
- `helpers.process_stopped` - Addition useful for checking a process is `SIGSTOP`'d

#### `feat: ruok livenessProbe for container health`
- Uses the `ruok` 4lw command to ping the local ZooKeeper server
- Defaults to `period=10s` and `threshold=3` as per the [Pebble specification](https://github.com/canonical/pebble#layer-specification)
- If fails, container is restarted as per [K8s specification](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-command)
#### `chore: bump juju client version`
- CI was failing without it, likely Juju 3.x defaulting
#### `feat: add cluster self-healing on pod reschedule`

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -51,7 +51,6 @@ jobs:
         tox-environments:
           - integration-password-rotation
           - integration-provider
-          - integration-scaling
```
Comment (deusebio) [inline .github/workflows/ci.yaml:54]: **question** Why have we remove this ? The file is still in the folder BTW

File: tox.ini [hunk 2]
```diff
@@ -21,6 +21,12 @@ set_env =
     PYTHONPATH = {tox_root}/lib:{[vars]src_path}
     PYTHONBREAKPOINT=ipdb.set_trace
     PY_COLORS=1
+    provider: TEST_FILE=test_provider.py
+    password-rotation: TEST_FILE=test_password_rotation.py
+    tls: TEST_FILE=test_tls.py
+    upgrade: TEST_FILE=test_upgrade.py
+    ha: TEST_FILE=ha/test_ha.py
+    replication: TEST_FILE=ha/test_replication.py
```
Comment (deusebio) [inline tox.ini:29]: **todo** I'm not sure this file exists. Why is this here?

Comment (deusebio) [reply tox.ini:29]: **question** also...why did we remove the scaling?

**Review by deusebio — COMMENTED** [review]
Just a couple of questions, but I believe the PR is almost ready to be merged

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (deusebio) [reply .github/workflows/ci.yaml:54]: All right. Got my answer, seen the others PR

File: tests/integration/ha/test_ha.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+    )
+    total_writes_leader = cw.count_znodes(
+        parent=parent, hosts=new_leader_host, username=helpers.USERNAME, password=password
+    )
+    assert last_write == last_write_leader
+    assert total_writes == total_writes_leader
+
+
+@pytest.mark.abort_on_fail
+async def test_kill_db_process(ops_test: OpsTest, request):
```
Comment (zmraul) [inline tests/integration/ha/test_ha.py:94]: **question**: Isn't this one missing the pebble restart delay?

File: tests/integration/ha/test_ha.py [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+async def test_full_cluster_crash(ops_test: OpsTest, request):
+    hosts = helpers.get_hosts(ops_test)
+    leader_name = helpers.get_leader_name(ops_test, hosts)
+    leader_host = helpers.get_unit_host(ops_test, leader_name)
+    password = helpers.get_super_password(ops_test)
+    parent = request.node.name
+    non_leader_hosts = ",".join([host for host in hosts.split(",") if host != leader_host])
+
+    logger.info("Extending pebble restart delay on all units...")
+    helpers.modify_pebble_restart_delay(ops_test, policy="extend")
```
Comment (zmraul) [inline tests/integration/ha/test_ha.py:302]: **suggestion**: I would keep this as a fixture, same as the [`restart_delay` here](https://github.com/canonical/zookeeper-operator/blob/main/tests/integration/ha/test_ha.py#L26)

File: tests/integration/ha/test_ha.py [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+
+    logger.info("Checking old leader caught up...")
+    last_write_leader = cw.get_last_znode(
+        parent=parent, hosts=new_leader_host, username=helpers.USERNAME, password=password
+    )
+    total_writes_leader = cw.count_znodes(
+        parent=parent, hosts=new_leader_host, username=helpers.USERNAME, password=password
+    )
+    assert last_write == last_write_leader
+    assert total_writes == total_writes_leader
```
Comment (zmraul) [inline tests/integration/ha/test_ha.py:90]: **todo**: Not on this PR, but we definitely need to extract all this on a different function. It's a lot of repeated code on each test. Some house cleaning for both charms.

**Review by zmraul — APPROVED** [review]
Great! Left some comments

File: tests/integration/ha/test_ha.py (hunk 4 above)
Comment (marcoppenheimer) [reply tests/integration/ha/test_ha.py:302]: Agreed, in another PR though.

File: tests/integration/ha/test_ha.py (hunk 3 above)
Comment (marcoppenheimer) [reply tests/integration/ha/test_ha.py:94]: We don't need it here I don't think, as we're hoping it actually restarts straight away anyway.

File: tests/integration/ha/test_ha.py (hunk 5 above)
Comment (marcoppenheimer) [reply tests/integration/ha/test_ha.py:90]: Agreed!

File: ; [hunk 6]
```diff
@@ -0,0 +1,528 @@
+import logging
```
Comment (zmraul) [inline ;:1]: ?

**Review by zmraul — CHANGES_REQUESTED** [review]

File: ; (hunk 6 above)
Comment (marcoppenheimer) [reply ;:1]: Lmao how'd that get there.

File: tests/integration/ha/test_ha.py (hunk 3 above)
Comment (zmraul) [reply tests/integration/ha/test_ha.py:94]: All other _kill_db_process_ methods **do** have a restart delay, as the intention is to check that writes still continue when the leader went down and we have less remaining units.

**Review by zmraul — APPROVED** [review]
